Add Search Dropdown

Accio1 opened the pull request:
### Resolves:

Resolves #257 

### Changes:

-Adds a search dropdown
-The search dropdown will display search suggestions based on what a user has typed into the search bar.
-Only 5 suggestions will be shown at once, so typing in "a" will display the first 5 users with "a" in their name.
-Case insensitive
-Clicking on a suggestion will input that user into the search bar.

### Local Tests:

Tested locally. Feel free to test and see if you can find a way to break it.

The original issue didn't have much discussion, so feel free to suggest changes/additions here. I will leave this PR open for a few days so everyone has the chance to suggest changes.

@leahcimto @jeffalo @gosoccerboy5

gosoccerboy5 commented on forumhelpers/FHP.js:
```diff
@@ -66,8 +68,33 @@ function getCount(name) {
+	searchBar.value = fillValue;
+	checkUser();
+	updateUserChoice();
+}
```
On click of the autofill value hopefully should automatically scroll down to the user.

gosoccerboy5 replied:
Skimming the code shows that _maybe_ you already intended to do that, but it's not working for me. Otherwise, great PR! :heart:

Accio1 replied:
Good idea! It should be as easy as running searchUser() under updateUserChoice() in the code you selected above.

gosoccerboy5 commented on forumhelpers/FHP.js:
```diff
@@ -6,6 +7,7 @@ function getData(dataSet) {
 			document.getElementById(dataSet+"Number").innerHTML = data.length;
 			for (j = 0; j < data.length; j++) {
 				var name = data[j].name;
+				membersList.push(name)
```
a few places in the code, like this, have no semicolons as line terminators. is it possible to fix that?

Accio1 replied:
Yep, I can.

gosoccerboy5 commented:
And while you're at it, could you support using `enter` key (I believe the keycode is 13) to submit the username, instead of forcing the user to click the button?

Accio1 commented:
I just realised that I might have misunderstood what @jeffalo suggested in [this](https://github.com/theforumhelpers/theforumhelpers.github.io/issues/257#issuecomment-869782858) comment. I think he might have meant that it hides the actual users on the page, so that only ones matching search criteria show up. I will see what I can to to implement this as an option tomorrow.

Accio1 commented:
@gosoccerboy5 I just pushed the changes you suggested.

Accio1 commented:
Converting to draft pending the change that I mentioned above to make this have the functionality that jeffalo suggested in the original issue.

gosoccerboy5 commented:
> Converting to draft pending the change that I mentioned above to make this have the functionality that jeffalo suggested in the original issue.


in the checkUser() function, I added:
```js
	for (let element of document.querySelectorAll(".profileContainer")) {
		if (element.id.includes(searchedUser)) {
			element.style.display = "";
		} else {
			element.style.display = "none";
		}
	}
```
and I changed the addition variable in the getData function to:
```js
`
				<div class="profileContainer" id="${name.toUpperCase()}">
					<h4 class="profileName"><a href="https://scratch.mit.edu/users/${name}/">${name}</a><span id="${name}Posts"></span></h4>
					<div class="profileInner">
						<a href="https://scratch.mit.edu/users/${name}/"><img src="https://cdn2.scratch.mit.edu/get_image/user/${id}_60x60.png" class="profilePicture" id="${name}pfp" loading="lazy" alt="${name}'s Profile Picture"></a>
						<p class="profileBio">${bio}</p>
					</div>
					<p class="ocularStatusBox" id="${name}Ocular" title="Ocular Status"><span class="ocularStatus" id="${name}Status">Loading Status...</span></p>
					<br>
					<hr>
				</div>`
```
(I had to move the `<hr>` element inside the `<div>` because otherwise the `<hr>` element keeps showing even though the users and their statuses/bios hide)

Accio1 commented:
> > Converting to draft pending the change that I mentioned above to make this have the functionality that jeffalo suggested in the original issue.
> 
> in the checkUser() function, I added:
> 
> ```js
> 	for (let element of document.querySelectorAll(".profileContainer")) {
> 		if (element.id.includes(searchedUser)) {
> 			element.style.display = "";
> 		} else {
> 			element.style.display = "none";
> 		}
> 	}
> ```
> 
> and I changed the addition variable in the getData function to:
> 
> ```js
> `
> 				<div class="profileContainer" id="${name.toUpperCase()}">
> 					<h4 class="profileName"><a href="https://scratch.mit.edu/users/${name}/">${name}</a><span id="${name}Posts"></span></h4>
> 					<div class="profileInner">
> 						<a href="https://scratch.mit.edu/users/${name}/"><img src="https://cdn2.scratch.mit.edu/get_image/user/${id}_60x60.png" class="profilePicture" id="${name}pfp" loading="lazy" alt="${name}'s Profile Picture"></a>
> 						<p class="profileBio">${bio}</p>
> 					</div>
> 					<p class="ocularStatusBox" id="${name}Ocular" title="Ocular Status"><span class="ocularStatus" id="${name}Status">Loading Status...</span></p>
> 					<br>
> 					<hr>
> 				</div>`
> ```
> 
> (I had to move the `<hr>` element inside the `<div>` because otherwise the `<hr>` element keeps showing even though the users and their statuses/bios hide)

I'm working on something similar. I think I am going to make it an option on whether the search bar will use the original functionality that this PR implemented or the functionality that jeffalo had in mind. I'll add a checkbox next to the search that allows the user to toggle between the two options.

gosoccerboy5 commented:
Okay, I made a new branch. https://gosoccerboy5.github.io/theforumhelpers.github.io/forumhelpers/ is currently checked out on that branch so you can directly view my changes

Accio1 commented:
Marking as ready for review because the most recent commit adds a checkbox which gives the user the option of the dropdown search, or removing users that don't match search criteria. This is what I think jeffalo meant in [this](https://github.com/theforumhelpers/theforumhelpers.github.io/issues/257#issuecomment-869782858) comment.

@leahcimto @gosoccerboy5 @jeffalo

gosoccerboy5 reviewed (approved):
Looks great!! :rocket:
One last thing - 
<img width="700" alt="Screen Shot 2021-06-29 at 2 46 11 PM" src="https://user-images.githubusercontent.com/82768218/123858127-c4298600-d8e8-11eb-9cef-c52ee380ca33.png">
Is it possible to make the search suggestion hide if it is exactly equal to the input value? It's a tiny bit annoying

leahcimto commented on forumhelpers/index.html:
```diff
@@ -43,20 +43,25 @@ <h3 class="forumHelpersHeader">Below is a list of current Forum Helpers, along w
+			<input type="text" id="searchBar" placeholder="Username" oninput="checkUser(), updateUserChoice()" onkeydown="if (event.keyCode == 13); document.getElementById('searchButton').click()">
 			<button id="searchButton" onclick="searchUser()" disabled title="Invalid Username">Search</button>
+			<div id="searchOptions"></div>
+			<br>
```
Could you add another `<br>` tag right under this one (or make the line height x2 bigger) since search bar and the checkbox look a bit close together.

(Also, not sure what others feel about this but I kind of feel like there shouldn't be a checkbox and it should just hide matching results by default since I'm not sure why someone would want to see the whole list and just a dropdown when searching that has the possible results, but that is for another issue and isn't apart of this changes requested thing)

jeffalo commented:
why is it a dropdown? wouldn't it be clearer if it just automatically hid the things that don't match the query?

(eg. if i search jef, the forum helper list should only show people with names starting with jef)

i guess one advantage of the dropdown is it still lets you see the other forum helpers, but i dont know if thats really important.

if i'm searching for something, i'm expecting to see my results straight away instead of having a drop down that scrolls me down to the right place.

or maybe im just a bit weird :P

Accio1 commented:
I think that having both options is good, because then if someone wants to just see matching search results, they can, or if they want to be able to jump to a specific user within the entire list, they can as well. Perhaps the checkbox value should be saved in localStorage so that the user's selection is remembered when they next visit the site?

Accio1 commented:
Whoops, I missed the thumbs up reactions to my post above. I will update this now to reflect the change that @leahcimto suggested above, as well as storing the user's selection in localstorage.

Accio1 commented:
@leahcimto @jeffalo I just made the changes discussed above. Sorry for the delay!

Accio1 commented:
Merging!

gosoccerboy5 commented:
do you think a squash merge would be possible? i guess it's only 5 commits, but it would keep the commit history a bit neater

Accio1 commented:
Alright, good to merge now.

Accio1 commented:
> do you think a squash merge would be possible? i guess it's only 5 commits, but it would keep the commit history a bit neater

@gosoccerboy5 I always Squash and Merge.
